src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -460,16 +460,38 @@ class Typer extends Namer with TypeAssigner with Applications with Implicits wit
   }
 
   def typedAssign(tree: untpd.Assign, pt: Type)(implicit ctx: Context) = track("typedAssign") {
+    def curriedUpdate(canCopy:untpd.Tree,update:untpd.Tree,lhsArgs:List[untpd.Tree]) = {
+      val step1 = cpy.Apply(canCopy)(update,lhsArgs)
+      typed(tree.rhs match {
+        case untpd.Tuple(rhsArgs) => untpd.Apply(step1,rhsArgs)
+        case rhsNotATuple => untpd.Apply(step1,rhsNotATuple)
+      },pt)
+    }
+    /** method tries normal update syntax, if that causes a type error, switch to curried mode
+      * not an effcient way to do this...
+      * Uses uncurried method in some ambiguous cases; should maybe give an error/warning.
+      */
+    def compatibilityMode(canCopy:untpd.Tree,update:untpd.Tree,lhsArgs:List[untpd.Tree]) = {
+      val trying = ctx.fresh.setNewTyperState
+      val uncurriedUpdate = cpy.Apply(canCopy)(update, lhsArgs :+ tree.rhs)(trying)
```

Comment:
There's a tryEither method in the Typer that embodies this logic. It seems DRYer to use that.